Pr gazebo (#499)
ros-planning/moveit_tutorials: doc/gazebo_simulation/gazebo_simulation.rst

----------------------------
After MoveIt Setup Assistant
----------------------------
This tutorial assumes that the robot is set up with `MoveIt Setup Assistant <../setup_assistant/setup_assistant_tutorial.html>`_,
so it is crucial to follow that document first. To the best of our knowledge, official FRANKA EMIKA Panda repository doesn't particularly consider Gazebo simulation
such that the necessary components to properly simulate the robot in Gazebo are missing with respect to :code:`urdf` and :code:`xacro` files. This is a rare incident, since most other robots
have those components out of the box. *If you have a custom robot, which already works well in Gazebo, you can skip the steps until Step-6.* Fortunately, there is already a good solution offered in `the blog post <https://erdalpekel.de/?p=55>`_ to this problem. For the sake of completion though,
the procedure outlined in there will be repeated (with improvements) in here as well for preparing the robot for Gazebo simulation.

> **Note that these steps assume that you have cloned franka_ros repository from the source**:

Reviewer: It would be helpful to include a link to the repository.

Author: It assumes that the user had followed the Setup Assistant tutorial, which has that link.

Reviewer: ```suggestion
**Note that these steps assume that you have cloned the `franka_ros repository <https://github.com/frankaemika/franka_ros>` from the source**:
```

Reviewer: It does have the link, but it also recommends installing the package via `apt-get`, so I didn't actually need the link until I got to this step, which is why I suggest adding the link.

1. Fix the robot to the world coordinate system
2. Add damping to the joint specifications
3. Add inertia matrices and masses to the links
4. Add friction and colorize the links
5. Configure gazebo_ros_control, transmissions and actuators
6. Adjust auto-generated ros_controllers.yaml
7. Adjust auto-generated ros_controllers.launch

1. Fix the robot to the world coordinate system
-----------------------------------------------
Open the :code:`franka_description/robots/panda_arm_hand_urdf.xacro` file and change the line 5 with:

.. code-block:: xml

    <xacro:panda_arm xyz="0 0 0" rpy="0 0 0" connected_to="world"/>

It alone doesn't fix the problem, since now we need to provide a link with name :code:`world`. Open the 
:code:`franka_description/robots/panda_arm.xacro` file and add:

.. code-block:: xml

>     <link name="world" />

Reviewer 2: I would suggest to add the `<link name="world" />` inside `franka_description/robots/panda_arm_hand_urdf.xacro` and not in `franka_description/robots/panda_arm.xacro`. I think is more clean since `panda_arm_hand_urdf.xacro` call the macro specifying `connect_to="world"` and therefore I expect the world to be defined outside the macro.

between lines 3 and 4. Additionally, we should rename the fixed joint to :code:`virtual_joint` to properly match

> the SRDF specification created in previous tutorial.

Reviewer: Is this on line 8? It probably is, but it'd be very helpful to be more specific.

    </inertial>

Similarly add following snippet to the :code:`${ns}_hand` link in :code:`franka_description/robots/hand.xacro` file:

.. code-block:: xml

    <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="0.68"/>
        <inertia ixx="0.1" ixy="0.0" ixz="0.0" iyy="0.1" iyz="0.0" izz="0.1"/>
    </inertial>

Finally add following inertia block to :code:`finger` links:

.. code-block:: xml

    <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="0.01"/>
        <inertia ixx="0.1" ixy="0.0" ixz="0.0" iyy="0.1" iyz="0.0" izz="0.1"/>
    </inertial>

As previously mentioned, these values come from the referred blog post. It is explicitly advised to have a look in there to grasp the matter in-depth.

4. Add friction and colorize the links
--------------------------------------
In order to have a nice illustration of the robot in Gazebo simulation we need to colorize the links. 
Moreover friction forces are added in order to have realistic dynamics. You can ignore them at all or change their values to experiment.
Since the focus is MoveIt in this tutorial, we will just use the values from the provided solution.

> This step is a bit tedious to do manually, so the ultimate :code:`xacro` file is provided entirely in below:

Reviewer: Where does this file go? What should it be called?

.. code-block:: xml

    <!-- panda.gazebo.xacro -->
    <?xml version='1.0' encoding='utf-8'?>

>     <robot xmlns:xacro="http://www.ros.org/wiki/xacro">

Reviewer 2: By a blind copy of the code-block  (as I did) you get an error during the urdf generation from xacro file.
Xacro file have to start with `<?xml version='1.0' encoding='utf-8'?>`. 
I suggest therefore to swap line 1 with line 2. In this way we still have the information on the file name which is very useful, but the copy and paste will work.

                    <material>Gazebo/White</material>
                    <mu1>0.2</mu1>
                    <mu2>0.2</mu2>
                </gazebo>
            </xacro:macro>
            <xacro:macro name="hand_gazebo" params="link">
                <gazebo reference="${link}">
                    <material>Gazebo/Grey</material>
                    <mu1>0.2</mu1>
                    <mu2>0.2</mu2>
                </gazebo>
            </xacro:macro>
            <xacro:arm_gazebo link="${arm_id}_link0"/>
            <xacro:arm_gazebo link="${arm_id}_link1"/>
            <xacro:arm_gazebo link="${arm_id}_link2"/>
            <xacro:arm_gazebo link="${arm_id}_link3"/>
            <xacro:arm_gazebo link="${arm_id}_link4"/>
            <xacro:arm_gazebo link="${arm_id}_link5"/>
            <xacro:arm_gazebo link="${arm_id}_link6"/>
            <xacro:hand_gazebo link="${arm_id}_link7"/>
            <xacro:hand_gazebo link="${arm_id}_link8"/>
            <xacro:hand_gazebo link="${arm_id}_hand"/>
            <xacro:hand_gazebo link="${arm_id}_rightfinger"/>
            <xacro:hand_gazebo link="${arm_id}_leftfinger"/>
        </xacro:macro>
    </robot>

> Then add the following block to the end of :code:`franka_description/robots/panda_arm_hand_urdf.xacro` file:

Reviewer: Before the `</robot>`?

Author: Yeah. I mean, come on :sweat_smile:

.. code-block:: xml

    <xacro:include filename="$(find franka_description)/robots/panda.gazebo.xacro"/>
    <xacro:panda_gazebo arm_id="panda"/>

5. Configure gazebo_ros_control, transmissions and actuators
------------------------------------------------------------

This is necessary for the robot to move in Gazebo. ROS Control is a highly capable robot-agnostic stack, providing interfaces
to control theoretically any type of robot. :code:`gazebo_ros_control` enables the ROS control to be used in Gazebo. 
See `its document <http://gazebosim.org/tutorials/?tut=ros_control>`_ for full details.

> Add the following to the newly created :code:`panda.gazebo.xacro` file:

Reviewer: Where in the file? Inside the `<robot>` tag?

Author: Just before `</robot>`. I assumed some level of intimacy..

Reviewer 3: It's a tutorial that many first-time users will read though, and it's easy to trip up and get lost on "that should be obvious" things. Better to err on the side of too much detail. Don't forget that if you can write a tutorial in the first place, you automatically tend to underestimate the difficulty.

.. code-block:: xml

    <gazebo>
        <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so"/>
    </gazebo>

> For transmissions and actuators we are going to create a new file, :code:`panda.control.xacro`. As before, I will provide the full content now:

Reviewer: Where does this file go?

Reviewer 2: I suggest to add the `libgazebo_ros_contro.so` in a separate xacro file.
Inserting it in the `panda.gazebo.xacro` will not allow to use multiple time the `panda_gazebo` macro since we will instantiate multiple gazebo_ros_control plugin that can interfere one with the other. Having a separate macro to add the gazebo_ros_control plugin in my experience is an effective solution. So when you are building a robot you can call several time the `panda_gazebo` macro with different arm_id and at the end you only add once the gazebo_plugin

Author: Good point @ValerioMa :+1:

                </transmission>
            </xacro:macro>
            <xacro:arm_control transmission="${arm_id}_tran_1" joint="${arm_id}_joint1" motor="${arm_id}_motor_1"/>
            <xacro:arm_control transmission="${arm_id}_tran_2" joint="${arm_id}_joint2" motor="${arm_id}_motor_2"/>
            <xacro:arm_control transmission="${arm_id}_tran_3" joint="${arm_id}_joint3" motor="${arm_id}_motor_3"/>
            <xacro:arm_control transmission="${arm_id}_tran_4" joint="${arm_id}_joint4" motor="${arm_id}_motor_4"/>
            <xacro:arm_control transmission="${arm_id}_tran_5" joint="${arm_id}_joint5" motor="${arm_id}_motor_5"/>
            <xacro:arm_control transmission="${arm_id}_tran_6" joint="${arm_id}_joint6" motor="${arm_id}_motor_6"/>
            <xacro:arm_control transmission="${arm_id}_tran_7" joint="${arm_id}_joint7" motor="${arm_id}_motor_7"/>
            <xacro:arm_control transmission="${arm_id}_leftfinger" joint="${arm_id}_finger_joint1" motor="${arm_id}_finger_joint1"/>
            <xacro:arm_control transmission="${arm_id}_rightfinger" joint="${arm_id}_finger_joint2" motor="${arm_id}_finger_joint2"/>
        </xacro:macro>
    </robot>

> Similarly add the following line to the end of :code:`franka_description/robots/panda_arm_hand_urdf.xacro` file:

Reviewer: Again, before `</robot>`, correct? Please specify.

Author: I really think _end of_ means just before `</robot>` in this context.

        joints:
            - panda_joint1
            - panda_joint2
            - panda_joint3
            - panda_joint4
            - panda_joint5
            - panda_joint6
            - panda_joint7
        gains:
            panda_joint1: { p: 12000, d: 50, i: 0.0, i_clamp: 10000 }
            panda_joint2: { p: 30000, d: 100, i: 0.02, i_clamp: 10000 }
            panda_joint3: { p: 18000, d: 50, i: 0.01, i_clamp: 1 }
            panda_joint4: { p: 18000, d: 70, i: 0.01, i_clamp: 10000 }
            panda_joint5: { p: 12000, d: 70, i: 0.01, i_clamp: 1 }
            panda_joint6: { p: 7000, d: 50, i: 0.01, i_clamp: 1 }
            panda_joint7: { p: 2000, d: 20, i: 0.0, i_clamp: 1 }

        constraints:
            goal_time: 2.0
        state_publish_rate: 25

    panda_hand_controller:
        type: effort_controllers/JointTrajectoryController
        joints:
            - panda_finger_joint1
            - panda_finger_joint2

        gains:
            panda_finger_joint1: { p: 5, d: 3.0, i: 0, i_clamp: 1 }
            panda_finger_joint2: { p: 5, d: 1.0, i: 0, i_clamp: 1 }

        state_publish_rate: 25

    controller_list:
        - name: panda_arm_controller

>             action_ns: follow_joint_trajectory

Reviewer 2: ```suggestion
          action_ns: follow_joint_trajectory
```

>             type: FollowJointTrajectory

Reviewer 2: ```suggestion
          type: FollowJointTrajectory
```

>             default: true

Reviewer 2: ```suggestion
          default: true
```

>             joints:

Reviewer 2: ```suggestion
          joints:
```

            - panda_joint1
            - panda_joint2
            - panda_joint3
            - panda_joint4
            - panda_joint5
            - panda_joint6
            - panda_joint7
        - name: panda_hand_controller
            action_ns: follow_joint_trajectory
            type: FollowJointTrajectory
            default: true
            joints:
            - panda_finger_joint1
            - panda_finger_joint2

> 

Reviewer 2: _action_ns_, _type_, _default_,  _joints_  should have the same indentation of _name_ both in _panda_arm_controller_ and _panda_hand_controller_   otherwise by just copy and pasting users will get the following error:

> ros_controllers.yaml contains invalid YAML:
mapping values are not allowed here
  in "<string>", line 63, column 18:
            action_ns: follow_joint_trajectory
                     ^
XML is <rosparam command="load" file="$(find panda_moveit_config)/config/ros_controllers.yaml"/>
The traceback for the exception was written to the log file

Author: Copy-pasting stuff between implicitly indented text editors (with their own rules, 2? 4? Or a tab, damn) is a long-lasting pain-in-the-neck for me :man_facepalming: Thanks for pointing it out :+1:

7. Adjust auto-generated ros_controllers.launch
-----------------------------------------------

> Fill the :code:`args` in line 9 with:

Reviewer: It'd be helpful to say exactly where to find the `ros_controllers.launch` file

Author: Fair :+1:

.. code-block:: xml

    joint_state_controller panda_hand_controller panda_arm_controller

> -------------------------------

Reviewer 2: I had to modify the _gazebo.launch_ to load the xacro file. The default launch file generated from MoveIt Setup Assistant was trying to load the xacro as urdf. Maybe it was mandatory to generate the URDF from the Setup Assistant? 
![Screenshot from 2020-07-16 16-22-23](https://user-images.githubusercontent.com/18074378/87682819-98b59600-c780-11ea-9de6-f232e71a8b45.png)

Author: We don't touch to the URDF in the setup assistant tutorial. In that tutorial, we create SRDF files that are used particularly by MoveIt.

You are right though, the auto-generated file has the line:
`<param name="robot_description" textfile="$(arg urdf_path)" /`  to load the robot description to parameter server. I forgot to mention this. Thanks for pointing out!

At the end, both repositories will have the updated and directly usable versions.

-------------------------------

Now it is time to integrate MoveIt to this work. Open :code:`panda_moveit_config/launch/demo_gazebo.launch` file 
and replace line 61 with:

.. code-block:: xml

    <arg name="rviz_config" value="$(find panda_moveit_config)/launch/moveit.rviz"/>

This will allow us to use juicy Motion Planning display of MoveIt in Rviz. There is a final minor issue in `demo_gazebo.launch` file. Remove
line 31 from that file, which contains unused :code:`urdf_path` argument. After that, launch:

.. code-block:: xml

    roslaunch panda_moveit_config demo_gazebo.launch

.. figure:: moveit-gazebo.gif
   :width: 700px

   Panda arm controlled via MoveIt in Gazebo simulation.

We have successfully integrated MoveIt and Gazebo ultimately. MoveIt Setup Assistant already does
many work under the hood, but it still misses some parts to provide a proper Gazebo integration. After following
this tutorial you should be able to reproduce this locally for any robot. In case you don't want to be 

> bothered with all the details, `franka_ros <https://github.com/tahsinkose/franka_ros>`_ and `panda_moveit_config <https://github.com/tahsinkose/panda_moveit_gazebo_config>`_

Reviewer: The link to `https://github.com/tahsinkose/panda_moveit_gazebo_config` is broken--maybe you meant `https://github.com/tahsinkose/panda_moveit_config`? Also, please say which branches should be used, or link directly to them. Looks like the `simulation` branch on `franka_ros` and the `melodic-devel` branch on `panda_moveit_config`, right?

Author: > The link to https://github.com/tahsinkose/panda_moveit_gazebo_config is broken--maybe you meant https://github.com/tahsinkose/panda_moveit_config?

Yes, thanks for notification :+1: 

> Looks like the simulation branch on franka_ros and the melodic-devel branch on panda_moveit_config, right

Yes to both.

Reviewer: ```suggestion
bothered with all the details, `franka_ros <https://github.com/tahsinkose/franka_ros/tree/simulation>`_ and `panda_moveit_config <https://github.com/tahsinkose/panda_moveit_config/tree/melodic-devel>`_
```

Author: Thanks!
